RFCs/FS-1060-nullable-reference-types.md

```diff
@@ -0,0 +1,1033 @@
+
+#### Likely unsupported: boolean-based checks for null that lack an appropriate well-known attribute
+
+Generally speaking, beyond highly-specialized cases, we cannot guarantee non-nullability that is checked via a boolean. Consider the following:
+
+```fsharp
+let myIsNull item = isNull item
+
+let len (str: string | null) =
+    if myIsNull str then
+        -1
+    else
+        str.Length // WARNING: could be null
+```
+
+Although this trivial example could certainly be done, it could quickly get out of hand. Also, other languages that support nullable reference types (C# 8.0, Scala, Kotlin, Swift, TypeScript) do not do this. An annotation for `myIsNull` is needed, otherwise a programmer will have to rewrite their code or assert non-nullability and risk a `NullReferenceException`.
+
+#### Likely unsupported: nested functions accessing outer scopes
+
+```fsharp
+let len (str: string | null) =
+    let doWork() =
+        str.Length // WARNING: maybe too complicated? Doesn't feel natural to F#?
+
+    match str with
+    | null -> -1
+    | _ -> doWork() // But after this line is written it's fine? Not natural to F# though.
+```
+
+Although `doWork()` is called only in a scope where `str` would be non-null, this may too complex to implement properly.
+
+**Note:** C# supports the equivalent of this with local functions. TypeScript does not support this. There is no precedent for this kind of support to further motivate the work.
+
+#### Asserting non-nullability
+
+To get around scenarios where compiler analysis cannot establish a non-null situation, a programmer can use the `notNull` function to convert a nullable reference type to a non-nullable reference type:
+
+```fsharp
+let len (ns: string | null) =
+    let s = notNull(ns) // unsafe, but lets you avoid a warning
+    s.Length
+```
+
+This function has a signature of `('T | null -> 'T)`. It will throw a `NullReferenceException` at runtime if the input is `null`.
+
+This sort of feature is inherently unsafe, and is by definition not something being enthusiastically considered. But it will be a necessity for when the compiler cannot analyze code well enough.
+
+#### Warnings
+
+In all other situations not covered by compiler analysis, a warning is emitted by the compiler if a nullable reference is dereferenced or cast to a non-null type:
+
+```fsharp
+let f (ns: string | null) =
+    printfn "%d" ns.Length // WARNING: 'ns' may be null
+
+    let mutable ns2 = ns
+    printfn "%d" ns2.Length // WARNING: 'ns2' may be null
+
+    let s = ns :> string // WARNING: 'ns' may be null
+```
+
+A warning is given when casting from `'S[]` to `('T | null)[]` and from `('S | null)[]` to `'T[]`.
+
+A warning is given when casting from `C<'S>` to `C<'T | null>` and from `C<'S | null>` to `C<'T>`.
+
+### Checking of non-null references
+
+#### Null assignment and passing
+
+A warning is given if `null` is assigned to a non-null value or passed as a parameter where a non-null reference type is expected:
```

Comment:
Do you mean: ... is assigned to a non-null ~~value~~ __variable__ ...?

Reply:
Also, aside from (1) assignment, and (2) argument-passing, there probably also is third case for returning. Or is that part of the "assignment" category?

Reply:
Re: `value` -> `variable` - we use `value` to refer to many things in F#, distinguishing between a value and a mutable value (with immutable the default).

Re: return value - I've added another example in the code sample that shows that if you attempt to assume the result of a function that returns a nullable reference type to a value that is explicitly non-nullable, it's a warning.